### PR TITLE
Fix separate installation of implementations

### LINKED ISSUE #2322: separate installation of virtual-modules packages fails
I've ported io-page to use virtual_modules, and it works great in a duniverse. However, when publishing separate opam packages and using `dune build -p`, they can no longer be installed since the source code is in different repos.  This is a pretty big blocker to using the feature, unless I'm doing something wrong in how I'm using it.

```
#=== ERROR while compiling io-page-xen.dev ====================================#
# context              2.0.4 | linux/x86_64 | ocaml-base-compiler.4.08.0 | pinned(file:///repo)
# path                 ~/.opam/4.08/.opam-switch/build/io-page-xen.dev
# command              /usr/bin/env OPAM_PKG_CONFIG_PATH=/home/opam/.opam/4.08/lib/pkgconfig dune build -p io-page-xen -j 71
# exit-code            1
# env-file             ~/.opam/log/io-page-xen-3173-be7d52.env
# output-file          ~/.opam/log/io-page-xen-3173-be7d52.out
### output ###
# File "lib/xen/dune", line 7, characters 14-21:
# 7 |   (implements io-page)
#                   ^^^^^^^
# Error: Library implementation io-page for variant "xen" implements a library outside the project. This is forbidden.
#=== ERROR while compiling io-page-unix.dev ===================================#
# context              2.0.4 | linux/x86_64 | ocaml-base-compiler.4.08.0 | pinned(file:///repo)
# path                 ~/.opam/4.08/.opam-switch/build/io-page-unix.dev
# command              ~/.opam/4.08/bin/dune build -p io-page-unix -j 71
# exit-code            1
# env-file             ~/.opam/log/io-page-unix-3173-a63268.env
# output-file          ~/.opam/log/io-page-unix-3173-a63268.out
### output ###
# File "lib/unix/dune", line 7, characters 14-21:
# 7 |   (implements io-page)
#                   ^^^^^^^
# Error: Library implementation io-page for variant "unix" implements a library outside the project. This is forbidden.
```

Travis log:
https://travis-ci.org/mirage/io-page/jobs/550969705

PR for io-page that shows this problem:
https://github.com/mirage/io-page/pull/59

### PR DESCRIPTION
**Pull request by TheLortex**
This fixes #2322

When instantiating an implementation with a variant, make sure it matches vlib's
list of known implementations. It replaces the same-project check that was done when
instantiating an implementation.

### CODE REVIEW

File: src/lib.ml [hunk 1]
```diff
@@ -864,48 +887,39 @@ let rec instantiate db name info ~stack ~hidden =
...
+        let name = Lib_info.name info in
+        let error () =
+          let vlib_name = Lib_info.name vlib.info in
+          Error.vlib_known_implementation_mismatch
+            ~loc ~name ~variant ~vlib_name
+        in
+        let vlib_impls = Lib_info.known_implementations vlib.info in
+        let* (_, impl_name) =
+          match Variant.Map.find vlib_impls variant with
+          | None -> error ()
```
Comment (rgrinberg) [inline src/lib.ml]: I've refactored the code a bit, and now it seems that the same error is used for both cases. Do you think we should perhaps use more descriptive errors for both cases?

**Comment by rgrinberg** [comment]
A slightly unrelated point that I raised on slack was the presence of this new `known_implementations` field in `Lib_info`. Would it be possible to move this field outside of `Lib_info`? It seems like a package property to me, so I think it's best stored in `Dune_package.t`? `Lib_info.t` is meant for library information only that is independent of the scope.

File: src/lib.ml (hunk 1 above)
Comment (TheLortex) [reply src/lib.ml]: Indeed, we can separate the errors to make them more specific:
* one case is when the variant is used for another library
* the other is when the library doesn't know about the variant

**Comment by TheLortex** [comment]
@rgrinberg rebased and added a second error message according to your review.

With regards to moving `known_implementations` to `Dune_package.t`, this seems sensible: should I try to accomplish that in another PR ?

**Comment by rgrinberg** [comment]
> With regards to moving known_implementations to Dune_package.t, this seems sensible: should I try to accomplish that in another PR ?

Sounds good to me. Thanks for improving the errors.

**Comment by rgrinberg** [comment]
Merged.

**Comment by avsm** [comment]
and confirmed to work in https://github.com/mirage/io-page/pull/59 -- thanks!

**Pull request closed without merging**